src/components/legend/draw.js

```diff
@@ -520,52 +570,116 @@ function handleClick(g, gd, numClicks) {
 
         Plotly.relayout(gd, 'hiddenlabels', hiddenSlices);
     } else {
-        var allTraces = [],
-            traceVisibility = [],
-            i;
-
-        for(i = 0; i < fullData.length; i++) {
-            allTraces.push(i);
-            // Allow the legendonly state through for *all* trace types (including
-            // carpet for which it's overridden with true/false in supplyDefaults)
-            traceVisibility.push(
-                Registry.traceIs(fullData[i], 'notLegendIsolatable') ? true : 'legendonly'
-            );
-        }
-
-        if(legendgroup === '') {
-            traceIndicesInGroup = [trace.index];
-            traceVisibility[trace.index] = true;
-        } else {
+        var hasLegendgroup = legendgroup && legendgroup.length;
+        var traceIndicesInGroup = [];
+        var tracei;
+        if(hasLegendgroup) {
             for(i = 0; i < fullData.length; i++) {
                 tracei = fullData[i];
+                if(!tracei.visible) continue;
                 if(tracei.legendgroup === legendgroup) {
-                    traceIndicesInGroup.push(tracei.index);
-                    traceVisibility[allTraces.indexOf(i)] = true;
+                    traceIndicesInGroup.push(i);
                 }
             }
         }
 
         if(numClicks === 1) {
-            newVisible = trace.visible === true ? 'legendonly' : true;
-            Plotly.restyle(gd, 'visible', newVisible, traceIndicesInGroup);
+            var nextVisibility;
+
+            switch(fullTrace.visible) {
+                case true:
+                    nextVisibility = 'legendonly';
+                    break;
+                case false:
+                    nextVisibility = false;
+                    break;
+                default:
+                case 'legendonly':
+                    nextVisibility = true;
+                    break;
+            }
+
+            if(hasLegendgroup) {
+                for(i = 0; i < fullData.length; i++) {
+                    if(fullData[i].visible && fullData[i].legendgroup === legendgroup) {
+                        setVisibility(fullData[i], nextVisibility);
+                    }
+                }
+            } else {
+                setVisibility(fullTrace, nextVisibility);
+            }
         } else if(numClicks === 2) {
-            var sameAsLast = true;
+            // Compute the clicked index. expandedIndex does what we want for expanded traces
+            // but also culls hidden traces. That means we have some work to do.
+            var clickedIndex, isIsolated, isClicked, isInGroup, otherState;
+            for(clickedIndex = 0; clickedIndex < fullData.length; clickedIndex++) {
```

Comment:
Isn't `clickedIndex` simply `fullTrace.index`?

Reply:
Let's see… it might be. I need to think about this just a bit. `fullTrace.index` vs. `fullTrace._fullInput.index` vs. `fullTrace._expandedIndex` obey slightly different rules. I think more often that not the answer was to avoid at all costs needing the index at all. I'll give the above a try.

Reply:
Ah, good catch. `clickedIndex` isn't actually used at all, which was the end result of trying to get this right (i.e. don't use it at all). I've removed this code.